doc/man/man3/XcmDDC.3
.TH "X Color Management DDC/CI communication API's" 3 "Tue May 26 2015" "Version 0.5.4" "Xcm-0.5.4" \" -*- nroff -*-
.ad l
.nh
.SH NAME
X Color Management DDC/CI communication API's \- 
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBXCM_DDC_ERROR_e\fP { \fBXCM_DDC_OK\fP, \fBXCM_DDC_NO_FILE\fP, \fBXCM_DDC_PREPARE_FAIL\fP, \fBXCM_DDC_EDID_READ_FAIL\fP, \fBXCM_DDC_WRONG_EDID\fP }"
.br
.RI "\fIerrors specific to XcmDDC \fP"
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "\fBXCM_DDC_ERROR_e\fP \fBXcmDDClist\fP (char ***list, int *count)"
.br
.RI "\fIobtain all monitors i2c device names \fP"
.ti -1c
.RI "\fBXCM_DDC_ERROR_e\fP \fBXcmDDCgetEDID\fP (const char *device_name, char **data, size_t *size)"
.br
.RI "\fIobtain all monitors i2c device names \fP"
.ti -1c
.RI "const char * \fBXcmDDCErrorToString\fP (\fBXCM_DDC_ERROR_e\fP error)"
.br
.RI "\fIconvert enum into a meaningful text string \fP"
.in -1c
.SH "Detailed Description"
.PP 
The purpose of the API is to establish client side communication with a monitor device\&. This can be used to update EEDID data and get MCCS properties\&. The former is supported by this library\&.
.PP
\fBXcmDDClist()\fP returns a list of monitor i2c device names\&. \fBXcmDDCgetEDID()\fP can be used to fetch EEDID from a device name\&.
.PP
XcmDDC\&.h
.PP
Xcm Xorg Colour Management
.PP
DDC communication with monitor devices
.PP
\fBCopyright:\fP
.RS 4
2010 (C) Kai-Uwe Behrmann
.RE
.PP
\fBAuthor:\fP
.RS 4
Kai-Uwe Behrmann ku.b@gmx.de 
.RE
.PP
\fBLicense:\fP
.RS 4
MIT http://www.opensource.org/licenses/mit-license.php 
.RE
.PP
\fBSince:\fP
.RS 4
2010/12/28 
.RE
.PP

.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBXCM_DDC_ERROR_e\fP"

.PP
errors specific to XcmDDC 
.PP
\fBVersion:\fP
.RS 4
libXcm: 0\&.4\&.0 
.RE
.PP
\fBSince:\fP
.RS 4
2010/12/28 (libXcm: 0\&.4\&.0) 
.RE
.PP
\fBDate:\fP
.RS 4
2010/12/28 
.RE
.PP

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIXCM_DDC_OK \fP\fP
success 
.TP
\fB\fIXCM_DDC_NO_FILE \fP\fP
fopen failed, permissions? 
.TP
\fB\fIXCM_DDC_PREPARE_FAIL \fP\fP
setting EEDID adress failed 
.TP
\fB\fIXCM_DDC_EDID_READ_FAIL \fP\fP
could not obtain EEDID 
.TP
\fB\fIXCM_DDC_WRONG_EDID \fP\fP
wrong EDID header 
.SH "Function Documentation"
.PP 
.SS "const char * XcmDDCErrorToString (\fBXCM_DDC_ERROR_e\fPerror)"

.PP
convert enum into a meaningful text string Function XcmDDCErrorToString 
.PP
\fBParameters:\fP
.RS 4
\fIerror\fP the error 
.RE
.PP
\fBReturns:\fP
.RS 4
library owned error text string
.RE
.PP
\fBVersion:\fP
.RS 4
libXcm: 0\&.4\&.0 
.RE
.PP
\fBSince:\fP
.RS 4
2010/12/28 (libXcm: 0\&.4\&.0) 
.RE
.PP
\fBDate:\fP
.RS 4
2010/12/28 
.RE
.PP

.PP
References XCM_DDC_EDID_READ_FAIL, XCM_DDC_NO_FILE, XCM_DDC_OK, XCM_DDC_PREPARE_FAIL, and XCM_DDC_WRONG_EDID\&.
.SS "\fBXCM_DDC_ERROR_e\fP XcmDDCgetEDID (const char *device_name, char **data, size_t *size)"

.PP
obtain all monitors i2c device names Function XcmDDCgetEDID 
.PP
\fBParameters:\fP
.RS 4
\fIdevice_name\fP system specific device name 
.br
\fIdata\fP empty pointer to EDID data, will be mallocd for the user 
.br
\fIsize\fP size of memory in data 
.RE
.PP
\fBReturns:\fP
.RS 4
error code
.RE
.PP
\fBVersion:\fP
.RS 4
libXcm: 0\&.4\&.0 
.RE
.PP
\fBSince:\fP
.RS 4
2010/12/28 (libXcm: 0\&.4,0) 
.RE
.PP
\fBDate:\fP
.RS 4
2010/12/28 
.RE
.PP

.PP
References XCM_DDC_NO_FILE, XCM_DDC_OK, XCM_DDC_PREPARE_FAIL, and XCM_DDC_WRONG_EDID\&.
.PP
Referenced by XcmDDClist()\&.
.SS "\fBXCM_DDC_ERROR_e\fP XcmDDClist (char ***list, int *count)"

.PP
obtain all monitors i2c device names Function XcmDDClist 
.PP
\fBParameters:\fP
.RS 4
\fIlist\fP empty pointer to a list of device names 
.br
\fIcount\fP pass in a pointer to a int\&. gives the number of elements in list 
.RE
.PP
\fBReturns:\fP
.RS 4
error code
.RE
.PP
\fBVersion:\fP
.RS 4
libXcm: 0\&.4\&.0 
.RE
.PP
\fBSince:\fP
.RS 4
2010/12/28 (libXcm: 0\&.4\&.0) 
.RE
.PP
\fBDate:\fP
.RS 4
2010/12/28 
.RE
.PP

.PP
References XCM_DDC_NO_FILE, XCM_DDC_OK, and XcmDDCgetEDID()\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for Xcm-0\&.5\&.4 from the source code\&.


ChangeLog
Version libXcm-0.5.3

Florian Höch (1):
	* [core]: Patch for transposed EDID red y bits

Kai-Uwe Behrmann (20):
	* [build]: remove all libtool files from RPM
	* [core]: fix conditional compiling of XcmDummy
	* [build]: fix spec packaging issues
	* [core]: call split libraries in libXcm
	* [build]: sync spec file with obs
	* [core]: add XcmDummy.c file
	* [build]: add xcm-edid xcm-x11 xcm-ddc pkg-config files
	* [build]: reorder building of libraries
	* [build]: split EDID parsing library out
	* [core]: XcmeContext_InLoop return meaningful value
	* [core]: Fix window name print
	* [conf]: update debian changelog
	* [docu]: update man pages
	* [core]: omit XmuClientWindow()
	* [conf]: bump version
	* [core]: initialy observe windows
	* [docu]: fix link
	* [build]: update autotools files
	* [build]: remove libtool
	* [build]: add missed automake script

Version libXcm-0.5.2

Jan Engelhardt (2):
	build: support for automake-1.12
	build: do not fail if AM_SILENT_RULES is not present

Jean-Sébastien Pédron (2):
	* [build]: support build dir != source dir
	* [build]: regen autotools files to correctly support FreeBSD 10

Kai-Uwe Behrmann (11):
	* [build]: rename FindXcm.cmake to XcmConfig.cmake
	* [conf]: install FindXcm.cmake
	* [core]: define XCM_COLOR_SERVER_MANAGEMENT enum
	* [docs]: describe ICM in X Color Management spec
	* [build]: fix debian install file names
	* [build]: add needed build requirements for debian
	* [build]: add requred dummy file
	* [build]: add make debian target
	* [docu]: mention opt out in X Color Management spec
	* [conf]: bump version
	* [build]: reset release version in RPM spec file

Version libXcm-0.5.1

Kai-Uwe Behrmann (4):
	* [core]: add XCM_ prefix to HAVE macros
	* [core]: simplify varg allocation for messages
	* [build]: depend on dist-bzip2 for make rpm target
	* [conf]: fix DDC code linking

Michael Weber (1):
	* [conf]: add --disable-x11 configure option


Version libXcm-0.5.0

Jan Engelhardt (7):
	* [conf]: allow building without X11
	* [build]: allow building without X11
	* [core]: resolve compiler warnings about incorrect code
	* [conf]: Xcm header files utilize header files from xproto (Xatom.h) and x11 (Xlib.h)
	* [core]: resolve compiler warnings about incorrect code
	* [conf]: switch to autotools
	* [conf]: Xcm header files utilize header files from xproto (Xatom.h) and x11 (Xlib.h)

Kai-Uwe Behrmann (75):
	* [build]: correct configure version macor in spec
	* [build]: update spec file for autotools changes
	* [core]: keep capabilities visible for non X11
	* [build]: add missed autotools files
	* [core]: skip "WARNING" for system message in xcmevents
	* [conf]: add xproto x11 conditionally to xcm.pc
	* [conf]: HAVE_X11 conditional for Xcm.h
	* [conf]: fix XcmEvents building
	* [conf]: copy build system
	* [core]: add XCM_COLOR_OUTPUTS to header
	* [docu]: adapt the X Color Management spec intro
	* [docu]: X Color Management specification 0.4
	* [conf]: fix Version field
	* [conf]: add Version field to pkg-config file
	* [conf]: fill all needed paths for pkg-config
	* [conf]: add config.h.in
	* [conf]: add build-aux files
	* [conf]: track essential build scripts and files
	* [conf]: remove old configuration files
	* [conf]: add status at end of configure run
	* [build]: add example files to package
	* [build]: add make help target
	* [build]: add make rpm and deb targets
	* [docu]: fix doxygen warning
	* [core]: use XcmVersion.h macros
	* [conf]: write HAVE_X11 to XcmVersion.h
	* [conf]: set fine grained package and library verions
	* [build]: pass libtool version to linker
	* [build]: readd missed files to tar ball
	* [conf]: adapt spec file template to autotools
	* [docu]: show macros for doxygen man pages
	* [conf]: add version header template
	* [conf]: add urls to configure.ac
	* [conf]: prepare autotools patch
	* [conf]: generalise Win32 detection
	* [conf]: check for MINGW32
	* [conf]: always use -fPIC
	* [core]: print ser_id
	* [core]: fix ser_id field to 32bit
	* [core]: add XcmColorServerCapabilities()
	* [core]: fix colorimetry name macros
	* [core]: fix string buffer allocation
	* [core]: fix key name swapping
	* [docu]: X Color Management v0.3 DRAFT2
	* [core]: adapt new XCM_COLOUR_DESKTOP_ADVANCED macro
	* [core]: detect _ICC_COLOR_DISPLAY_ADVANCED events
	* [docu]: add _ICC_COLOR_DISPLAY_ADVANCED + _ICC_COLOR_PROFILES
	* [core]: add a name to the XcmEvents observer window
	* [core]: add XcmeContext_Setup2 API
	* [core]: add SEC manufacturer
	* [core]: add GCM and LGE
	* [core]: be more strict about EDID primariy order
	* [core]: add LGD manufacturer
	* [conf]: bump to 0.5
	* [docu]: update to new XCM macro
	* [build]: fix spec file
	* [core]: use XCM_COLOR_ macros
	* [core]: device XCM atom strings
	* [docu]: correct the capabilities tagging XCM docu
	* [core]: rename net-color spec to X_Color_Management
	* [core]: export numbers as strings for JSON
	* [docu]: update autogenerated man pages
	* [exmpl]: use XcolorRegionDelete for 0 0 0 0 geometry
	* [core]: properly export XcmePrintWindowName
	* [build]: add client example make check target
	* [exmpl]: make profile selectable
	* [build]: add build line for color
	*  [exmpl]: compile color with C
	* [exmpl]: fix client
	* [build]: adapt spec file to openSUSE conventions
	* [exmpl]: add very simple makefile for color example
	* [core]: return Xorg error values
	* [exmpl]: update to actual library
	* [conf]: bump version
	* [core]: chose first text block for type

Version libXcm-0.4.2

Kai-Uwe Behrmann (11):
	* [core]: use only JSON array in xcmedid
	* [core]: skip the numbered JSON device level in xcmedid
	* [build]: default to bz2 compression for RPM
	* [build]: remove unused macros in RPM spec file
	* [core]: merge EDID week and year to EDID_date
	* [core]: fall back for EDID_model to model_id
	* [docu]: update XcmEdidParse man page
	* [core]: add AUO and LEN manufacturers
	* [core]: add XcmEdidPrintOpenIccJSON() API
	* [core]: exclude DDC from __FreeBSD__
	* [conf]: bump version


src/XcmDDC.c
/*  @file XcmDDC.c
 *
 *  libXcm  Xorg Colour Management
 *
 *  @par Copyright:
 *            2010 (C) Kai-Uwe Behrmann
 *
 *  @brief    DDC communication with monitor devices 
 *  @internal
 *  @author   Kai-Uwe Behrmann <ku.b@gmx.de>
 *  @par License:
 *            MIT <http://www.opensource.org/licenses/mit-license.php>
 *  @since    2010/12/28
 */

#include "XcmVersion.h"

#if XCM_HAVE_LINUX

#include "XcmDDC.h"

#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>       /* open(), write() */
#include <dirent.h>       /* opendir() */
#include <fcntl.h>        /* O_RDWR */
/*#include <sys/stat.h>*/
#include <unistd.h>       /* usleep() */

#include <sys/ioctl.h>
#include <linux/types.h>

#ifndef USE_GETTEXT
#define _(text) text
#endif

#ifndef I2C_SLAVE
#define I2C_SLAVE  0x0703
#endif

#ifndef I2C_SLAVE_FORCE
#define I2C_SLAVE_FORCE 0x0706
#endif

#define TIMEOUT 50000

#define I2C_DIR "/dev/"

/** \addtogroup XcmDDC X Color Management DDC/CI communication API's

 *  @{
 *
 *  The purpose of the API is to establish client side communication with a
 *  monitor device. This can be used to update EEDID data and get MCCS 
 *  properties. The former is supported by this library.
 *
 *  XcmDDClist() returns a list of monitor i2c device names.
 *  XcmDDCgetEDID() can be used to fetch EEDID from a device name.
 *
 */

/* basic access functions */

/** Function XcmDDClist
 *  @brief   obtain all monitors i2c device names
 *
 *  @param[out]    list                empty pointer to a list of device names
 *  @param[out]    count               pass in a pointer to a int. gives the 
 *                                     number of elements in list
 *  @return                            error code
 *
 *  @version libXcm: 0.4.0
 *  @since   2010/12/28 (libXcm: 0.4.0)
 *  @date    2010/12/28
 */
XCM_DDC_ERROR_e   XcmDDClist         ( char            *** list,
                                       int               * count )
{
  XCM_DDC_ERROR_e error = XCM_DDC_OK;

  DIR * dir = opendir(I2C_DIR);
  struct dirent * entry;
  char * data = 0;
  size_t size;
  char * fn;
  int n = 0;
  char ** devices;

  if(!dir)
    return XCM_DDC_NO_FILE;

  fn = calloc(sizeof(char),1024);
  if(!fn)
  {
    error = XCM_DDC_NO_FILE;
    goto clean_list;
  }

  devices = calloc(sizeof(char*), 128);
  if(!devices)
  {
    error = XCM_DDC_NO_FILE;
    goto clean_list;
  }

  do
  {
    entry = readdir( dir );
    if(entry && strstr( entry->d_name, "i2c-" ))
    {
      sprintf( fn, I2C_DIR "%s", entry->d_name );
      size = 0;
      error = XcmDDCgetEDID( fn, &data, &size );

      if(!error && (size % 128) == 0)
      {
        devices[n] = strdup(fn);
        ++n;
      }

      if(data)
        free(data);
      data = 0;
    }
  } while(entry);
 

  *count = n;

  if(n > 0)
  {
    *list = devices;
    devices = NULL;
  } else if(devices)
  {
    free(devices);
    devices = NULL;
  }

  clean_list:
  if(devices)
    free(devices);
  if(dir)
    closedir(dir);
  if(fn)
    free(fn);

  return error;
}


int fds[128];
int fd_n = -1;

/** Function XcmDDCgetEDID
 *  @brief   obtain all monitors i2c device names
 *
 *  @param[in]     device_name         system specific device name
 *  @param[out]    data                empty pointer to EDID data, will be 
 *                                     mallocd for the user
 *  @param[out]    size                size of memory in data
 *  @return                            error code
 *
 *  @version libXcm: 0.4.0
 *  @since   2010/12/28 (libXcm: 0.4,0)
 *  @date    2010/12/28
 */
XCM_DDC_ERROR_e   XcmDDCgetEDID      ( const char        * device_name,
                                       char             ** data,
                                       size_t            * size )
{
  int fd = 0,
      ret = 0;
  XCM_DDC_ERROR_e error = XCM_DDC_OK;
  char command[128] = {0};


  ++fd_n;

  fds[fd_n] = open( device_name, O_RDWR );
  fd = fds[fd_n];

  if(fd != -1)
  {
    /* request the first 128 byte EDID block */
    error = ioctl(fd, I2C_SLAVE, 0x50);
    if(error == XCM_DDC_OK)
    {
      if(TIMEOUT)
        usleep(TIMEOUT);
      ret = write( fd, &command, 1 );
    } else
      error = XCM_DDC_PREPARE_FAIL;

    if(ret == 1)
    {
      *data = (char*)calloc(sizeof(char),256);
      if(TIMEOUT)
        usleep(TIMEOUT);
      ret = read( fd, *data, 128 );
    } else
      error = XCM_DDC_PREPARE_FAIL;

    if((ret % 128) == 0)
    {
      unsigned char * edid = (unsigned char*) *data;

      if(
         edid[0] == 0 &&
         edid[1] == 255 &&
         edid[2] == 255 &&
         edid[3] == 255 &&
         edid[4] == 255 &&
         edid[5] == 255 &&
         edid[6] == 255 &&
         edid[7] == 0)
      {
        *size = 128;
      } else
        error = XCM_DDC_WRONG_EDID;
    }

    if(error == XCM_DDC_OK)
    {
      if(TIMEOUT)
        usleep(TIMEOUT);
      error = ioctl(fd, I2C_SLAVE_FORCE, 0x37);
      if(error != XCM_DDC_OK)
        error = XCM_DDC_PREPARE_FAIL;
    }
  } else
    error = XCM_DDC_NO_FILE;

  return error;
}


/** Function XcmDDCErrorToString
 *  @brief   convert enum into a meaningful text string
 *
 *  @param[in]     error               the error
 *  @return                            library owned error text string
 *
 *  @version libXcm: 0.4.0
 *  @since   2010/12/28 (libXcm: 0.4.0)
 *  @date    2010/12/28
 */
const char *   XcmDDCErrorToString   ( XCM_DDC_ERROR_e     error )
{
  const char * text = 0;
  switch(error)
  {
  case XCM_DDC_OK: text = ""; break;
  case XCM_DDC_NO_FILE: text = "Could not open file. Did you check permissions?."; break;
  case XCM_DDC_PREPARE_FAIL: text = "Setting EEDID adress failed."; break;
  case XCM_DDC_EDID_READ_FAIL: text = "Could not obtain EEDID."; break;
  case XCM_DDC_WRONG_EDID: text = "Wrong EDID header."; break;
  }
  return text;
}


/** @} XcmDDC */

#endif /* Linux */
